svcb.go

```diff
@@ -0,0 +1,759 @@
+	str := make([]string, len(s.Hint))
+	for i, e := range s.Hint {
+		x := e.To4()
+		if x == nil {
+			return "<nil>"
+		}
+		str[i] = x.String()
+	}
+	return strings.Join(str, ",")
+}
+
+func (s *SVCBIPv4Hint) parse(b string) error {
+	if strings.Contains(b, ":") {
+		return errors.New("dns: svcbipv4hint: expected ipv4, got ipv6")
+	}
+	str := strings.Split(b, ",")
+	dst := make([]net.IP, len(str))
+	for i, e := range str {
+		ip := net.ParseIP(e)
+		if ip == nil {
+			return errors.New("dns: svcbipv4hint: bad ip")
+		}
+		dst[i] = ip.To4()
+	}
+	s.Hint = dst
+	return nil
+}
+
+func (s *SVCBIPv4Hint) copy() SVCBKeyValue {
+	return &SVCBIPv4Hint{
+		append([]net.IP(nil), s.Hint...),
+	}
+}
+
+// SVCBECHConfig pair contains the ECHConfig structure defined in draft-ietf-tls-esni [RFC xxxx].
+// Basic use pattern for creating an echconfig option:
+//
+//	o := new(dns.HTTPS)
+//	o.Hdr.Name = "."
+//	o.Hdr.Rrtype = dns.HTTPS
+//	e := new(dns.SVCBECHConfig)
+//	e.ECH = "/wH...="
+//	o.Value = append(o.Value, e)
+type SVCBECHConfig struct {
+	ECH string // Base64 encoded ECHConfig
+}
+
+func (*SVCBECHConfig) Key() SVCBKey            { return SVCB_ECHCONFIG }
+func (s *SVCBECHConfig) copy() SVCBKeyValue    { return &SVCBECHConfig{s.ECH} }
+func (s *SVCBECHConfig) pack() ([]byte, error) { return []byte(s.ECH), nil }
+func (s *SVCBECHConfig) String() string        { return s.ECH }
+func (s *SVCBECHConfig) len() int              { return len(s.ECH) }
+
+func (s *SVCBECHConfig) unpack(b []byte) error {
+	s.ECH = string(b)
+	return nil
+}
+func (s *SVCBECHConfig) parse(b string) error {
+	s.ECH = b
+	return nil
+}
+
+// SVCBIPv6Hint pair suggests an IPv6 address which may be used to open connections
+// if A and AAAA record responses for SVCB's Target domain haven't been received.
+// In that case, optionally, A and AAAA requests can be made, after which the
+// connection to the hinted IP address may be terminated and a new connection may be opened.
+// Basic use pattern for creating an ipv6hint option:
+//
+//	o := new(dns.HTTPS)
+//	o.Hdr.Name = "."
+//	o.Hdr.Rrtype = dns.HTTPS
+//	e := new(dns.SVCBIPv6Hint)
+//	e.Hint = []net.IP{net.ParseIP("2001:db8::1")}
+//	o.Value = append(o.Value, e)
+type SVCBIPv6Hint struct {
+	Hint []net.IP
+}
+
+func (*SVCBIPv6Hint) Key() SVCBKey { return SVCB_IPV6HINT }
+func (s *SVCBIPv6Hint) len() int   { return 16 * len(s.Hint) }
+
+func (s *SVCBIPv6Hint) pack() ([]byte, error) {
+	b := make([]byte, 0, 16*len(s.Hint))
+	for _, e := range s.Hint {
+		if len(e) != net.IPv6len || e.To4() != nil {
+			return nil, errors.New("dns: svcbipv6hint: expected ipv6, hint is ipv4")
+		}
+		b = append(b, e...)
+	}
+	return b, nil
+}
+
+func (s *SVCBIPv6Hint) unpack(b []byte) error {
+	if len(b) == 0 || len(b)%16 != 0 {
+		return errors.New("dns: svcbipv6hint: ipv6 address byte array length not a multiple of 16")
+	}
+	x := make([]net.IP, 0, len(b)/16)
+	for i := 0; i < len(b); i += 16 {
+		x = append(x, net.IP(b[i:i+16]))
```

Comment:
This doesn’t check for IPv4 addresses, should it?

Reply:
It's impossible to check for the address type, this is just a byte array.

Reply:
A `To4() != nil` check was what I was referring to. Not sure if it’s a good idea, but it would seem consistent with other methods on this type.

Reply:
Well, if it's okay for you I would prefer not adding conditions that would always evaluate to one branch, `To4()` is always `nil` for 16-byte IP arrays.

Reply:
That seems inconsistent with other methods on this type that all have a `To4() != nil` check on them. It’s definitely not going to be a hotspot.

Reply:
OK, changed it.